src/realm/sync/network/default_socket.cpp

```diff
@@ -362,9 +365,153 @@ void DefaultWebSocketImpl::initiate_websocket_handshake()
+    , m_random{}
+    , m_user_agent{user_agent}
+    , m_state_mutex{}
+    , m_state{NotStarted}
+    , m_thread{}
+    , m_stop_future{}
+{
+    REALM_ASSERT(m_logger_ptr);                     // Make sure the logger is valid
+    util::seed_prng_nondeterministically(m_random); // Throws
+    start_keep_running_timer();                     // TODO: Update service so this timer is not needed
+}
+
+DefaultSocketProvider::~DefaultSocketProvider()
+{
+    // Wait for the thread to stop
+    stop(true);
+    if (m_keep_running_timer)
+        m_keep_running_timer->cancel();
+    if (m_thread.joinable()) {
+        m_thread.join();
+    }
+    REALM_ASSERT(m_state == Stopped || m_state == NotStarted);
+}
+
+bool DefaultSocketProvider::start()
+{
+    std::unique_lock<std::mutex> lock{m_state_mutex};
+    // Start the thread if it hasn't been started or is stopped
+    if (m_state != NotStarted && m_state != Stopped) {
+        return false;
+    }
+
+    // If the thread has been stopped, make sure it has been joined
+    if (m_state == Stopped) {
+        if (m_thread.joinable()) {
+            lock.unlock();
+            m_thread.join();
+            lock.lock();
+        }
+    }
+
+    m_logger_ptr->info("Event loop start()");
+    m_state = Started;
+    auto start_pf = util::make_promise_future<void>();
+    auto stop_pf = util::make_promise_future<void>();
+    m_stop_future = std::move(stop_pf.future);
+    m_thread = std::thread{
+        [this, start_promise = std::move(start_pf.promise), stop_promise = std::move(stop_pf.promise)]() mutable {
+            m_logger_ptr->info("Event loop thread running");
+            // The thread has started - release the main start() future
+            start_promise.emplace_value();
+            auto will_destroy_thread = util::make_scope_exit([&]() noexcept {
+                if (g_binding_callback_thread_observer) {
+                    g_binding_callback_thread_observer->will_destroy_thread();
+                }
+                thread_update_state(Stopped);
+                stop_promise.emplace_value();
+            });
+            if (g_binding_callback_thread_observer) {
+                g_binding_callback_thread_observer->did_create_thread();
+                try {
+                    {
+                        std::lock_guard<std::mutex> lock{m_state_mutex};
+                        // Was the thread stopped before we got here?
+                        if (m_state != Started) {
+                            m_logger_ptr->info("Event loop thread early exit");
+                            return;
+                        }
+                        m_state = Running;
+                    }
+                    m_service.run(); // Throws
+                }
+                catch (const std::exception& e) {
+                    m_logger_ptr->error("Event loop exception: ", e.what());
+                    g_binding_callback_thread_observer->handle_error(e);
+                    throw;
```

Comment:
In the old implementation we didn't re-throw the exception if the g_binding_callback_thread_observer was defined.